walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,42 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.sql.Time;
+import java.util.*;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    SortedSet<TimeRange> timeRangesSorted = new TreeSet<TimeRange>(TimeRange.ORDER_BY_START);
+    for(Event e: events){ //some map could be used here
+      if(hasAttendee(e, request.getAttendees())){
+        timeRangesSorted.add(e.getWhen());
+      }
+    }
+
+    TimeRange guardian = TimeRange.fromStartDuration(TimeRange.END_OF_DAY+1, 0); //Shouldn't timerange set outside the minutes of the day be invalid?
+    timeRangesSorted.add(guardian); //Could be done by an additional if at the end as well
+
+    int lastFinish = TimeRange.START_OF_DAY;
+    Collection<TimeRange> appropriateTimeranges = new ArrayList<TimeRange>();
+    for(TimeRange nextEvent: timeRangesSorted){
+      if(lastFinish < nextEvent.start()){ //should I extract nextEvent.start() to a local variable?
+        TimeRange empty = TimeRange.fromStartEnd(lastFinish, nextEvent.start(), false);
+        if(empty.duration() >= request.getDuration()){
+          appropriateTimeranges.add(empty);
+        }
+      }
+      if(lastFinish < nextEvent.end()){
+        lastFinish = nextEvent.end();
+      }
+    }
+
+    return appropriateTimeranges;
+  }
+
+  public boolean hasAttendee(Event event, Collection<String> Attendees){//will it delete the attendees if the event?
```

Comment:
1. Method can be private
2. attendees parameter should be lower cased - uppercase names are generally reserved for classes and interfaces, not instances of those.
3. The function could use some documentation (e.g. JavaDoc), explaining what it does (e.g. "Returns whether any of the attendees in the passed collection is participating in the event"). Reading the code allows the reader to comprehend what is the effect, but with short explanation it's easier to spot errors in implementation vs intention.